File: frontend/src/lib/analytics.ts
import { BASE_YEAR, CURRENT_MONTH, CURRENT_YEAR, INDICATORS, statusFromRatio } from "@/data/dataset";
import { DISTRICT_BY_ID, DISTRICTS } from "@/data/districts";
import { MODULES, MONTHS_SHORT } from "@/data/modules";
import type { Indicator, ModuleId, StatusId } from "@/lib/types";

export interface Filters {
  moduleId: ModuleId | "all";
  districtId: string | "all";
  year: number;
  status: StatusId | "all";
  period: "month" | "quarter" | "year";
}

export const DEFAULT_FILTERS: Filters = {
  moduleId: "all",
  districtId: "all",
  year: CURRENT_YEAR,
  status: "all",
  period: "year",
};

export function selectIndicators(f: Partial<Filters>): Indicator[] {
  return INDICATORS.filter(
    (i) =>
      (!f.moduleId || f.moduleId === "all" || i.moduleId === f.moduleId) &&
      (!f.districtId || f.districtId === "all" || i.districtId === f.districtId) &&
      (!f.year || i.year === f.year) &&
      (!f.status || f.status === "all" || i.status === f.status),
  );
}

export interface Rollup {
  plan: number;
  fact: number;
  /** fact/plan — 1.0 = reja bajarildi */
  ratio: number;
  /** Modul yo'nalishini hisobga olgan bajarilish (inflyatsiyada teskari) */
  performance: number;
  status: StatusId;
  count: number;
}

export function rollup(rows: Indicator[], moduleId?: ModuleId): Rollup {
  const plan = rows.reduce((s, r) => s + r.plan, 0);
  const fact = rows.reduce((s, r) => s + r.fact, 0);
  const isInflation = moduleId === "inflation" || rows.every((r) => r.moduleId === "inflation");
  // Inflyatsiya — foiz, yig'indi emas o'rtacha
  const p = isInflation && rows.length ? plan / rows.length : plan;
  const fx = isInflation && rows.length ? fact / rows.length : fact;
  const ratio = p === 0 ? 1 : fx / p;
  const lower = isInflation;
  const performance = lower ? (ratio === 0 ? 1 : 1 / ratio) : ratio;
  return {
    plan: p,
    fact: fx,
    ratio,
    performance,
    status: statusFromRatio(ratio, lower),
    count: rows.length,
  };
}

/** Har bir tuman uchun umumlashtirilgan bajarilish — xarita ranglari shundan. */
export function districtScores(
  moduleId: ModuleId | "all",
  year = CURRENT_YEAR,
): Record<string, Rollup> {
  const out: Record<string, Rollup> = {};
  for (const d of DISTRICTS) {
    if (moduleId === "all") {
      // Barcha modullar bo'yicha o'rtacha bajarilish
      const perModule = MODULES.map((m) =>
        rollup(
          INDICATORS.filter(
            (i) => i.districtId === d.id && i.moduleId === m.id && i.year === year,
          ),
          m.id,
        ),
      ).filter((r) => r.count > 0);
      const performance =
        perModule.reduce((s, r) => s + r.performance, 0) / Math.max(1, perModule.length);
      out[d.id] = {
        plan: perModule.reduce((s, r) => s + r.plan, 0),
        fact: perModule.reduce((s, r) => s + r.fact, 0),
        ratio: performance,
        performance,
        status: statusFromRatio(performance, false),
        count: perModule.reduce((s, r) => s + r.count, 0),
      };
    } else {
      out[d.id] = rollup(
        INDICATORS.filter(
          (i) => i.districtId === d.id && i.moduleId === moduleId && i.year === year,
        ),
        moduleId,
      );
    }
  }
  return out;
}

export interface MonthPoint {
  label: string;
  month: number;
  plan: number;
  fact: number;
  prev?: number;
}

/** Oylik dinamika: reja / amalda / o'tgan yil. */
export function monthlySeries(
  moduleId: ModuleId,
  districtId: string | "all",
  year = CURRENT_YEAR,
): MonthPoint[] {
  const isInflation = moduleId === "inflation";
  const points: MonthPoint[] = [];
  const lastMonth = year === CURRENT_YEAR ? CURRENT_MONTH : 12;

  for (let m = 1; m <= lastMonth; m++) {
    const pick = (y: number) =>
      INDICATORS.filter(
        (i) =>
          i.moduleId === moduleId &&
          i.year === y &&
          i.month === m &&
          (districtId === "all" || i.districtId === districtId),
      );
    const cur = pick(year);
    const prevRows = pick(year - 1);
    const agg = (rows: Indicator[], key: "plan" | "fact") => {
      if (!rows.length) return 0;
      const sum = rows.reduce((s, r) => s + r[key], 0);
      return round(isInflation ? sum / rows.length : sum);
    };
    points.push({
      label: MONTHS_SHORT[m - 1],
      month: m,
      plan: agg(cur, "plan"),
      fact: agg(cur, "fact"),
      prev: prevRows.length ? agg(prevRows, "fact") : undefined,
    });
  }
  return points;
}

export interface QuarterPoint {
  label: string;
  plan: number;
  fact: number;
}

export function quarterlySeries(
  moduleId: ModuleId,
  districtId: string | "all",
  year = CURRENT_YEAR,
): QuarterPoint[] {
  const isInflation = moduleId === "inflation";
  return [1, 2, 3, 4]
    .map((q) => {
      const rows = INDICATORS.filter(
        (i) =>
          i.moduleId === moduleId &&
          i.year === year &&
          i.quarter === q &&
          (districtId === "all" || i.districtId === districtId),
      );
      if (!rows.length) return null;
      const agg = (key: "plan" | "fact") => {
        const sum = rows.reduce((s, r) => s + r[key], 0);
        return round(isInflation ? sum / rows.length : sum);
      };
      return { label: `${q}-chorak`, plan: agg("plan"), fact: agg("fact") };
    })
    .filter(Boolean) as QuarterPoint[];
}

/** Yillik o'sish sur'ati (%) — joriy yil vs o'tgan yil, taqqoslanadigan oylar. */
export function yoyGrowth(moduleId: ModuleId, districtId: string | "all" = "all"): number {
  const upTo = CURRENT_MONTH;
  const sum = (year: number) => {
    const rows = INDICATORS.filter(
      (i) =>
        i.moduleId === moduleId &&
        i.year === year &&
        (i.month ?? 0) <= upTo &&
        (districtId === "all" || i.districtId === districtId),
    );
    if (!rows.length) return 0;
    const total = rows.reduce((s, r) => s + r.fact, 0);
    return moduleId === "inflation" ? total / rows.length : total;
  };
  const a = sum(BASE_YEAR);
  const b = sum(CURRENT_YEAR);
  if (!a) return 0;
  return round(((b - a) / a) * 100);
}

export interface WeakSpot {
  districtId: string;
  districtName: string;
  moduleId: ModuleId;
  moduleName: string;
  performance: number;
  gap: number;
  status: StatusId;
  note?: string;
}

/** Reja bajarilmayotgan tuman × modul kesimlari — eng og'iridan boshlab. */
export function weakSpots(
  year = CURRENT_YEAR,
  opts: { moduleId?: ModuleId | "all"; districtId?: string | "all"; limit?: number } = {},
): WeakSpot[] {
  const { moduleId = "all", districtId = "all", limit = 12 } = opts;
  const out: WeakSpot[] = [];

  for (const d of DISTRICTS) {
    if (districtId !== "all" && d.id !== districtId) continue;
    for (const m of MODULES) {
      if (moduleId !== "all" && m.id !== moduleId) continue;
      const rows = INDICATORS.filter(
        (i) => i.districtId === d.id && i.moduleId === m.id && i.year === year,
      );
      if (!rows.length) continue;
      const r = rollup(rows, m.id);
      if (r.performance >= 0.9) continue;
      const noted = rows.find((x) => x.note);
      out.push({
        districtId: d.id,
        districtName: d.name,
        moduleId: m.id,
        moduleName: m.name,
        performance: r.performance,
        gap: round(m.lowerIsBetter ? r.fact - r.plan : r.plan - r.fact),
        status: r.status,
        note: noted?.note,
      });
    }
  }
  return out.sort((a, b) => a.performance - b.performance).slice(0, limit);
}

/** Tuman kesimidagi reyting — modul bo'yicha bajarilish foizi. */
export function districtRanking(moduleId: ModuleId | "all", year = CURRENT_YEAR) {
  const scores = districtScores(moduleId, year);
  return DISTRICTS.map((d) => ({
    id: d.id,
    name: d.name,
    performance: scores[d.id]?.performance ?? 0,
    fact: scores[d.id]?.fact ?? 0,
    plan: scores[d.id]?.plan ?? 0,
    status: scores[d.id]?.status ?? ("critical" as StatusId),
  })).sort((a, b) => b.performance - a.performance);
}

/** Bitta tumanning barcha modullar bo'yicha kesimi — xarita tooltip/paneli uchun. */
export function districtProfile(districtId: string, year = CURRENT_YEAR) {
  const d = DISTRICT_BY_ID[districtId];
  const modules = MODULES.map((m) => {
    const rows = INDICATORS.filter(
      (i) => i.districtId === districtId && i.moduleId === m.id && i.year === year,
    );
    const r = rollup(rows, m.id);
    return {
      moduleId: m.id,
      name: m.short,
      unit: m.unit,
      color: m.color,
      plan: round(r.plan),
      fact: round(r.fact),
      performance: r.performance,
      status: r.status,
      yoy: yoyGrowth(m.id, districtId),
    };
  });
  const overall = modules.reduce((s, m) => s + m.performance, 0) / modules.length;
  return {
    district: d,
    modules,
    overall,
    status: statusFromRatio(overall, false),
    criticalCount: modules.filter((m) => m.status === "critical" || m.status === "at_risk").length,
  };
}

/** Umumiy KPI kartochkalari uchun. */
export function headlineStats(moduleId: ModuleId | "all", year = CURRENT_YEAR) {
  const mods = moduleId === "all" ? MODULES.map((m) => m.id) : [moduleId];
  const perf =
    mods
      .map((m) => rollup(INDICATORS.filter((i) => i.moduleId === m && i.year === year), m).performance)
      .reduce((s, v) => s + v, 0) / mods.length;

  const spots = weakSpots(year, { moduleId, limit: 200 });
  const critical = spots.filter((s) => s.status === "critical").length;
  const atRisk = spots.filter((s) => s.status === "at_risk").length;
  const growth =
    mods.map((m) => yoyGrowth(m)).reduce((s, v) => s + v, 0) / mods.length;

  return {
    performance: perf,
    growth: round(growth),
    critical,
    atRisk,
    districts: DISTRICTS.length,
    records: INDICATORS.filter(
      (i) => i.year === year && (moduleId === "all" || i.moduleId === moduleId),
    ).length,
  };
}

/** Modullar kesimidagi radar — tumanni respublika o'rtachasi bilan solishtirish. */
export function radarCompare(districtId: string | "all", year = CURRENT_YEAR) {
  return MODULES.map((m) => {
    const republic = rollup(INDICATORS.filter((i) => i.moduleId === m.id && i.year === year), m.id);
    const local =
      districtId === "all"
        ? republic
        : rollup(
            INDICATORS.filter(
              (i) => i.moduleId === m.id && i.districtId === districtId && i.year === year,
            ),
            m.id,
          );
    return {
      label: m.short,
      moduleId: m.id,
      local: round(Math.min(130, local.performance * 100)),
      republic: round(Math.min(130, republic.performance * 100)),
    };
  });
}

export function round(n: number, digits = 1): number {
  const f = 10 ** digits;
  return Math.round(n * f) / f;
}

File: backend/app/services/analytics.py
"""Bazadagi ko'rsatkichlar ustidan agregatsiyalar.

Frontend'dagi `src/lib/analytics.ts` bilan bir xil qoidalar — bir xil raqamlar
ikkala tomonda ham chiqishi uchun.
"""

from __future__ import annotations

from dataclasses import dataclass

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models import District, Indicator, Module

MONTHS_SHORT = [
    "Yan", "Fev", "Mar", "Apr", "May", "Iyn",
    "Iyl", "Avg", "Sen", "Okt", "Noy", "Dek",
]


def status_from_ratio(ratio: float, lower_is_better: bool) -> str:
    """fact/plan nisbatidan holat. Inflyatsiyada nisbat teskari o'qiladi."""
    r = (1 / ratio if ratio else 1.0) if lower_is_better else ratio
    if r >= 1.0:
        return "completed"
    if r >= 0.9:
        return "in_progress"
    if r >= 0.75:
        return "at_risk"
    return "critical"


@dataclass
class Rollup:
    plan: float
    fact: float
    ratio: float
    performance: float
    status: str
    count: int


def rollup(rows: list[Indicator], lower_is_better: bool) -> Rollup:
    if not rows:
        return Rollup(0.0, 0.0, 1.0, 1.0, "in_progress", 0)

    plan_sum = sum(r.plan for r in rows)
    fact_sum = sum(r.fact for r in rows)
    # Inflyatsiya — foiz, shuning uchun yig'indi emas o'rtacha
    plan = plan_sum / len(rows) if lower_is_better else plan_sum
    fact = fact_sum / len(rows) if lower_is_better else fact_sum

    ratio = fact / plan if plan else 1.0
    performance = (1 / ratio if ratio else 1.0) if lower_is_better else ratio
    return Rollup(
        plan=round(plan, 2),
        fact=round(fact, 2),
        ratio=round(ratio, 4),
        performance=round(performance, 4),
        status=status_from_ratio(ratio, lower_is_better),
        count=len(rows),
    )


def module_map(db: Session) -> dict[str, Module]:
    return {m.id: m for m in db.scalars(select(Module)).all()}


def district_map(db: Session) -> dict[str, District]:
    return {d.id: d for d in db.scalars(select(District)).all()}


def fetch(
    db: Session,
    *,
    module_id: str | None = None,
    district_id: str | None = None,
    year: int | None = None,
    max_month: int | None = None,
) -> list[Indicator]:
    stmt = select(Indicator)
    if module_id:
        stmt = stmt.where(Indicator.module_id == module_id)
    if district_id:
        stmt = stmt.where(Indicator.district_id == district_id)
    if year:
        stmt = stmt.where(Indicator.year == year)
    if max_month:
        stmt = stmt.where(Indicator.month <= max_month)
    return list(db.scalars(stmt).all())


def latest_year(db: Session) -> int:
    return db.scalar(select(func.max(Indicator.year))) or 0


def latest_month(db: Session, year: int) -> int:
    return db.scalar(select(func.max(Indicator.month)).where(Indicator.year == year)) or 12


def district_scores(db: Session, module_id: str | None, year: int) -> list[dict]:
    """Har bir tumanning bajarilish darajasi — xarita ranglari uchun."""
    mods = module_map(db)
    dists = district_map(db)
    rows = fetch(db, module_id=module_id, year=year)

    by_district: dict[str, list[Indicator]] = {}
    for r in rows:
        by_district.setdefault(r.district_id, []).append(r)

    out = []
    for did, d in dists.items():
        group = by_district.get(did, [])
        if module_id:
            agg = rollup(group, mods[module_id].lower_is_better)
            performance = agg.performance
            plan, fact = agg.plan, agg.fact
        else:
            # Barcha modullar bo'yicha o'rtacha bajarilish
            per_module = []
            for mid, m in mods.items():
                sub = [r for r in group if r.module_id == mid]
                if sub:
                    per_module.append(rollup(sub, m.lower_is_better))
            performance = (
                sum(p.performance for p in per_module) / len(per_module)
                if per_module
                else 1.0
            )
            plan = sum(p.plan for p in per_module)
            fact = sum(p.fact for p in per_module)

        out.append(
            {
                "district_id": did,
                "name": d.name,
                "plan": round(plan, 2),
                "fact": round(fact, 2),
                "performance": round(performance, 4),
                "status": status_from_ratio(performance, False),
            }
        )
    out.sort(key=lambda x: x["performance"], reverse=True)
    return out


def monthly_series(
    db: Session, module_id: str, district_id: str | None, year: int
) -> list[dict]:
    m = module_map(db)[module_id]
    rows = fetch(db, module_id=module_id, district_id=district_id, year=year)

    by_month: dict[int, list[Indicator]] = {}
    for r in rows:
        if r.month:
            by_month.setdefault(r.month, []).append(r)

    points = []
    for month in sorted(by_month):
        group = by_month[month]
        agg = rollup(group, m.lower_is_better)
        points.append(
            {
                "label": MONTHS_SHORT[month - 1],
                "plan": agg.plan,
                "fact": agg.fact,
            }
        )
    return points


def quarterly_series(
    db: Session, module_id: str, district_id: str | None, year: int
) -> list[dict]:
    m = module_map(db)[module_id]
    rows = fetch(db, module_id=module_id, district_id=district_id, year=year)

    by_q: dict[int, list[Indicator]] = {}
    for r in rows:
        if r.quarter:
            by_q.setdefault(r.quarter, []).append(r)

    out = []
    for q in sorted(by_q):
        agg = rollup(by_q[q], m.lower_is_better)
        out.append({"label": f"{q}-chorak", "plan": agg.plan, "fact": agg.fact})
    return out


def yoy_growth(db: Session, module_id: str, district_id: str | None = None) -> float:
    """Yillik o'sish (%). Taqqoslash faqat mavjud oylar bo'yicha."""
    cur = latest_year(db)
    if not cur:
        return 0.0
    prev = cur - 1
    up_to = latest_month(db, cur)
    m = module_map(db)[module_id]

    def total(y: int) -> float:
        rows = fetch(
            db, module_id=module_id, district_id=district_id, year=y, max_month=up_to
        )
        if not rows:
            return 0.0
        s = sum(r.fact for r in rows)
        return s / len(rows) if m.lower_is_better else s

    a, b = total(prev), total(cur)
    if not a:
        return 0.0
    return round((b - a) / a * 100, 2)


def weak_spots(
    db: Session,
    year: int,
    *,
    module_id: str | None = None,
    district_id: str | None = None,
    limit: int = 12,
) -> list[dict]:
    """Reja bajarilmayotgan tuman × soha kesimlari — eng og'iridan boshlab."""
    mods = module_map(db)
    dists = district_map(db)
    rows = fetch(db, module_id=module_id, district_id=district_id, year=year)

    grouped: dict[tuple[str, str], list[Indicator]] = {}
    for r in rows:
        grouped.setdefault((r.district_id, r.module_id), []).append(r)

    out = []
    for (did, mid), group in grouped.items():
        m = mods[mid]
        agg = rollup(group, m.lower_is_better)
        if agg.performance >= 0.9:
            continue
        noted = next((g.note for g in group if g.note), None)
        gap = agg.fact - agg.plan if m.lower_is_better else agg.plan - agg.fact
        out.append(
            {
                "district_id": did,
                "district_name": dists[did].name,
                "module_id": mid,
                "module_name": m.name,
                "performance": agg.performance,
                "gap": round(gap, 2),
                "status": agg.status,
                "note": noted,
            }
        )
    out.sort(key=lambda x: x["performance"])
    return out[:limit]


def district_profile(db: Session, district_id: str, year: int) -> dict:
    """Bir tumanning barcha sohalar kesimi."""
    mods = module_map(db)
    d = district_map(db)[district_id]
    rows = fetch(db, district_id=district_id, year=year)

    modules = []
    for mid, m in mods.items():
        sub = [r for r in rows if r.module_id == mid]
        agg = rollup(sub, m.lower_is_better)
        modules.append(
            {
                "module_id": mid,
                "name": m.short or m.name,
                "unit": m.unit,
                "plan": agg.plan,
                "fact": agg.fact,
                "performance": agg.performance,
                "status": agg.status,
                "yoy": yoy_growth(db, mid, district_id),
            }
        )

    overall = sum(x["performance"] for x in modules) / len(modules) if modules else 1.0
    return {
        "district": {
            "id": d.id,
            "name": d.name,
            "center": d.center,
            "population": d.population,
            "area_km2": d.area_km2,
        },
        "modules": modules,
        "overall": round(overall, 4),
        "status": status_from_ratio(overall, False),
    }


def headline_stats(db: Session, module_id: str | None, year: int) -> dict:
    mods = module_map(db)
    ids = [module_id] if module_id else list(mods)

    perfs = []
    for mid in ids:
        rows = fetch(db, module_id=mid, year=year)
        perfs.append(rollup(rows, mods[mid].lower_is_better).performance)

    spots = weak_spots(db, year, module_id=module_id, limit=500)
    return {
        "performance": round(sum(perfs) / len(perfs), 4) if perfs else 1.0,
        "growth": round(sum(yoy_growth(db, mid) for mid in ids) / len(ids), 2) if ids else 0.0,
        "critical": sum(1 for s in spots if s["status"] == "critical"),
        "at_risk": sum(1 for s in spots if s["status"] == "at_risk"),
        "records": len(fetch(db, module_id=module_id, year=year)),
    }
